Compile bug: undefined symbol: vkResetQueryPool for  android-ndk-r29-beta3

### Git commit

master branch bbbf5ec 

### Operating systems

Other? (Please let us know in description)

### GGML backends

Vulkan

### Problem description & steps to reproduce

There is no vkResetQueryPool for android vulkan.

```sh
[100%] Linking CXX executable ../../bin/llama-server
ld.lld: error: undefined symbol: vkResetQueryPool
>>> referenced by vulkan.hpp:2212 (/opt/d/C/vulkan/include/vulkan/vulkan.hpp:2212)
>>>               ../../ggml/src/ggml-vulkan/llama-server.lto.libggml-vulkan.a(ggml-vulkan.cpp.o at 83020).o:(ggml_backend_vk_graph_compute(ggml_backend*, ggml_cgraph*))
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [tools/server/CMakeFiles/llama-server.dir/build.make:118: bin/llama-server] Error 1
make[1]: *** [CMakeFiles/Makefile2:531: tools/server/CMakeFiles/llama-server.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

### First Bad Commit

_No response_

### Compile command

```shell
cmake -DGGML_CPU=ON -DGGML_CPU_AARCH64=ON -DGGML_CUDA_FA=OFF -DGGML_CUDA_GRAPHS=OFF -DGGML_HIP_NO_VMM=OFF -DGGML_HIP_MMQ_MFMA=OFF -DGGML_LASX=OFF -DGGML_LSX=OFF -DGGML_METAL=OFF -DGGML_METAL_EMBED_LIBRARY=OFF -DGGML_METAL_NDEBUG=OFF -DGGML_METAL_USE_BF16=OFF -DGGML_ACCELERATE=OFF -DGGML_OPENCL=OFF -DGGML_OPENCL_EMBED_KERNELS=OFF -DGGML_OPENCL_USE_ADRENO_KERNELS=OFF -DGGML_OPENMP=ON -DGGML_RVV=OFF -DGGML_VXE=OFF -DGGML_SYCL_GRAPH=OFF -DGGML_VULKAN=ON -DGGML_BLAS=OFF -DGGML_LTO=ON -DLLAMA_CURL=OFF -DBUILD_SHARED_LIBS=OFF -DLLAMA_BUILD_EXAMPLES=OFF -DLLAMA_BUILD_TESTS=OFF -DLLAMA_BUILD_SERVER=ON -DLLAMA_BUILD_TOOLS=ON -DGGML_LLAMAFILE=OFF
```

### Relevant log output

```shell
use https://github.com/KhronosGroup/Vulkan-Headers.git with android-ndk-r29-beta3, enable vulkan.
```

--- a/ggml/src/ggml-vulkan/ggml-vulkan.cpp
+++ b/ggml/src/ggml-vulkan/ggml-vulkan.cpp
@@ -5,7 +5,13 @@
 #include "ggml-cpu.h"
 #endif
 
+// See https://github.com/KhronosGroup/Vulkan-Hpp?tab=readme-ov-file#extensions--per-device-function-pointers-
+#define VULKAN_HPP_DISPATCH_LOADER_DYNAMIC 1
+
 #include <vulkan/vulkan.hpp>
+
+// See https://github.com/KhronosGroup/Vulkan-Hpp?tab=readme-ov-file#extensions--per-device-function-pointers-
+VULKAN_HPP_DEFAULT_DISPATCH_LOADER_DYNAMIC_STORAGE
 
 #include <algorithm>
 #include <cmath>
@@ -121,6 +127,8 @@ struct vk_pipeline_struct {
     bool needed {};
     // set to true when the shader has been compiled
     bool compiled {};
+    // number of registers used, extracted from pipeline executable properties
+    uint32_t register_count {};
 };
 
 typedef std::shared_ptr<vk_pipeline_struct> vk_pipeline;
@@ -428,6 +436,8 @@ struct vk_device_struct {
     uint32_t coopmat_int_k;
 
     bool coopmat2;
+
+    bool pipeline_executable_properties_support {};
 
     size_t idx;
 
@@ -1517,6 +1527,20 @@ static void ggml_vk_create_pipeline_func(vk_device& device, vk_pipeline& pipelin
         duoni.pObjectName = pipeline->name.c_str();
         duoni.objectHandle = /*reinterpret_cast*/(uint64_t)(static_cast<VkPipeline>(pipeline->pipeline));
         vk_instance.pfn_vkSetDebugUtilsObjectNameEXT(device->device, &static_cast<VkDebugUtilsObjectNameInfoEXT &>(duoni));
+    }
+
+    if (device->pipeline_executable_properties_support) {
+        vk::PipelineExecutableInfoKHR executableInfo;
+        executableInfo.pipeline = pipeline->pipeline;
+
+        auto statistics = device->device.getPipelineExecutableStatisticsKHR(executableInfo);
+        for (auto & s : statistics) {
+            // "Register Count" is reported by NVIDIA drivers.
+            if (strcmp(s.name, "Register Count") == 0) {
+                VK_LOG_DEBUG(pipeline->name << " " << s.name << ": " << s.value.u64 << " registers");
+                pipeline->register_count = (uint32_t)s.value.u64;
+            }
+        }
     }
 
     {
@@ -3509,6 +3533,7 @@ static vk_device ggml_vk_get_device(size_t idx) {
         bool amd_shader_core_properties2 = false;
         bool pipeline_robustness = false;
         bool coopmat2_support = false;
+        bool pipeline_executable_properties_support = false;
         device->coopmat_support = false;
         device->integer_dot_product = false;
         bool bfloat16_support = false;
@@ -3551,6 +3576,8 @@ static vk_device ggml_vk_get_device(size_t idx) {
                        !getenv("GGML_VK_DISABLE_BFLOAT16")) {
                 bfloat16_support = true;
 #endif
+            } else if (strcmp("VK_KHR_pipeline_executable_properties", properties.extensionName) == 0) {
+                pipeline_executable_properties_support = true;
             }
         }
 
@@ -3771,7 +3798,17 @@ static vk_device ggml_vk_get_device(size_t idx) {
             device_extensions.push_back("VK_KHR_shader_integer_dot_product");
         }
 
+        VkPhysicalDevicePipelineExecutablePropertiesFeaturesKHR pep_features {};
+        pep_features.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PIPELINE_EXECUTABLE_PROPERTIES_FEATURES_KHR;
+        if (pipeline_executable_properties_support) {
+            last_struct->pNext = (VkBaseOutStructure *)&pep_features;
+            last_struct = (VkBaseOutStructure *)&pep_features;
+            device_extensions.push_back("VK_KHR_pipeline_executable_properties");
+        }
+
         vkGetPhysicalDeviceFeatures2(device->physical_device, &device_features2);
+
+        device->pipeline_executable_properties_support = pipeline_executable_properties_support;
 
         device->fp16 = device->fp16 && vk12_features.shaderFloat16;
 
@@ -4288,6 +4325,9 @@ static void ggml_vk_instance_init() {
     }
     VK_LOG_DEBUG("ggml_vk_instance_init()");
 
+    // See https://github.com/KhronosGroup/Vulkan-Hpp?tab=readme-ov-file#extensions--per-device-function-pointers-
+    VULKAN_HPP_DEFAULT_DISPATCHER.init(vkGetInstanceProcAddr);
+
     uint32_t api_version = vk::enumerateInstanceVersion();
 
     if (api_version < VK_API_VERSION_1_2) {
@@ -4354,6 +4394,9 @@ static void ggml_vk_instance_init() {
     }
 
     vk_perf_logger_enabled = getenv("GGML_VK_PERF_LOGGER") != nullptr;
+
+    // See https://github.com/KhronosGroup/Vulkan-Hpp?tab=readme-ov-file#extensions--per-device-function-pointers-
+    VULKAN_HPP_DEFAULT_DISPATCHER.init(vk_instance.instance);
 
     std::vector<vk::PhysicalDevice> devices = vk_instance.instance.enumeratePhysicalDevices();
 

